File: LeetCode/practice-cpp/LeetCode-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// 两数相加
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode dummyHead = ListNode(0, nullptr);
        ListNode* curNode = &dummyHead;
        bool flag = flag; // 处理好进位

        while (l1 != nullptr || l2 != nullptr || flag) {
            ListNode* newNode = new ListNode(0);
            if (l1 != nullptr) {
                newNode->val += l1->val;
                l1 = l1->next;
            }
            if (l2 != nullptr) {
                newNode->val += l2->val;
                l2 = l2->next;
            }
            if (flag) {
                newNode->val += flag;
                flag = false;
            }
            
            if (newNode->val >= 10) {
                newNode->val -= 10;
                flag = true;
            }
            curNode->next = newNode;
            curNode = newNode;
        }
        return dummyHead.next;
    }
};

File: LeetCode/practice-rust/LeetCode-2.rs
// 两数相加
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>
}

impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode {
      next: None,
      val
    }
  }
}

struct Solution;

impl Solution {
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut dummy_head = Some(Box::new(ListNode::new(0)));
        let mut curr_node = &mut dummy_head;
        let mut carry = 0;
        let mut p1 = &l1;
        let mut p2 = &l2;

        while p1.is_some() || p2.is_some() || carry != 0 {
            let mut sum = carry;

            if let Some(node) = p1 {
                sum += node.val;
                p1 = &node.next;
            }

            if let Some(node) = p2 {
                sum += node.val;
                p2 = &node.next;
            }

            carry = sum / 10;
            sum %= 10;

            curr_node.as_mut().unwrap().next = Some(Box::new(ListNode::new(sum)));
            curr_node = &mut curr_node.as_mut().unwrap().next;

            // let next_node = Some(Box::new(ListNode::new(sum)));
            // if let Some(node) = curr_node {
            //     node.next = next_node;
            //     curr_node = &mut node.next;
            // }
        }
        dummy_head.unwrap().next
    }
}

fn main () {
// 测试用例: 342 + 465 = 807
    // l1: 2 -> 4 -> 3
    let l1 = Some(Box::new(ListNode {
        val: 2,
        next: Some(Box::new(ListNode {
            val: 4,
            next: Some(Box::new(ListNode::new(3))),
        })),
    }));
    // l2: 5 -> 6 -> 4
    let l2 = Some(Box::new(ListNode {
        val: 5,
        next: Some(Box::new(ListNode {
            val: 6,
            next: Some(Box::new(ListNode::new(4))),
        })),
    }));

    let result = Solution::add_two_numbers(l1, l2);
    println!("{:?}", result);
}